### PR TITLE
Feature/cdssm preprocessor

### PR DESCRIPTION
**Pull request by ZizhenWang**
#328

### CODE REVIEW

**Comment by ZizhenWang** [comment]
1. I use padding in `cdssm_preprocessor`, not `FixedLengthUnit`, because the padding value and axis are different, and it seems need a lot time refactoring the orignial unit to fit the specific cdssm model. So I write here
https://github.com/faneshion/MatchZoo/blob/2398559eb87a7190668a35212455171c79886b0d/matchzoo/preprocessor/cdssm_preprocessor.py#L136-L149
2. as 1 states, to handle `pad_value` in `cdssm_preprocessor`, which means to avoid `NgramLetterUnit` splitting it and `VocabularyUnit` indexing it, I states a `remove` list in these units, and it don't influence previous code
ngram
https://github.com/faneshion/MatchZoo/blob/2398559eb87a7190668a35212455171c79886b0d/matchzoo/preprocessor/process_units.py#L216-L218
vocabulary
https://github.com/faneshion/MatchZoo/blob/2398559eb87a7190668a35212455171c79886b0d/matchzoo/preprocessor/process_units.py#L287-L290

File: matchzoo/preprocessor/cdssm_preprocessor.py [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+        :param truncate_mode: truncate mode, `pre` or `post`.
+        :param remove: user-defined removed tokens.
+        """
+        self._datapack = None
+        if truncate_mode not in ['pre', 'post']:
+            raise ValueError('{} is not a vaild '
+                             'truncate mode.'.format(truncate_mode))
+        if pad_mode not in ['pre', 'post']:
+            raise ValueError('{} is not a vaild '
+                             'pad mode.'.format(pad_mode))
```
Comment (bwanglzu) [inline matchzoo/preprocessor/cdssm_preprocessor.py]: why copy code from fixed length unit?

File: matchzoo/preprocessor/cdssm_preprocessor.py [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+        ngram_unit = preprocessor.NgramLetterUnit(self._remove)
+
+        logger.info("Start building vocabulary & fitting parameters.")
+
+        # Convert user input into a datapack object.
+        self._datapack = self.segment(inputs, stage='train')
+
+        # Loop through user input to generate tri-letters.
+        # 1. Used for build vocabulary of tri-letters (get dimension).
+        # 2. Cached tri-letters can be further used to perform input
```
Comment (bwanglzu) [inline matchzoo/preprocessor/cdssm_preprocessor.py]: do not copy-paste comments

File: matchzoo/preprocessor/cdssm_preprocessor.py [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+                text = unit.transform(text)
+            fixed = np.full([self._length], self._pad_value)
+            if self._truncate_mode == 'pre':
+                trunc_text = text[-self._length:]
+            else:
+                trunc_text = text[:self._length]
+            if self._pad_mode == 'post':
+                fixed[:len(trunc_text)] = trunc_text
+            else:
+                fixed[-len(trunc_text):] = trunc_text
```
Comment (bwanglzu) [inline matchzoo/preprocessor/cdssm_preprocessor.py]: what are you doing here? from line 146-154

File: matchzoo/preprocessor/cdssm_preprocessor.py [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
```
Comment (bwanglzu) [inline matchzoo/preprocessor/cdssm_preprocessor.py]: again, do not copy-paste code, 164-172

**Pull request closed without merging**